[ONNX] Fix ignored small eps in layer normalization in fp16

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89869

Prior to this change, the symbolic_fn `layer_norm` (before ONNX version 17) always lose precision when eps is smaller than Float type, while PyTorch always take eps as Double. This PR adds `onnx::Cast` into eps related operations to prevent losing precision during the calculation.

titaiwangms committed:
[ONNX] Fix ignored small eps in fp16

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Fix ignored small eps in layer normalization in fp16"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Fix ignored small eps in layer normalization in fp16"


Prior to this change, the symbolic_fn `layer_norm` (before ONNX version 17) always lose precision when eps is smaller than Float type, while PyTorch always take eps as Double. This PR adds `onnx::Cast` into eps related operations to prevent losing precision during the calculation.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Fix ignored small eps in layer normalization in fp16"


Prior to this change, the symbolic_fn `layer_norm` (before ONNX version 17) always lose precision when eps is smaller than Float type, while PyTorch always take eps as Double. This PR adds `onnx::Cast` into eps related operations to prevent losing precision during the calculation.

[ghstack-poisoned]

BowenBao commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -2701,22 +2701,39 @@ def native_layer_norm(
     mean = g.op("ReduceMean", input, axes_i=axes)
     numerator = sub(g, input, mean)
+    # Cast it to eps dtype to avoid precision loss
+    cast_numerator = g.op(
```
Would it be necessary to cast all the time, or only when numerator is float16?

titaiwangms replied:
In torch, EPS is always Double, while in ONNX, we used to keep it Float32 all the time without any issue. This Cast would make us stay with Double all the time.

BowenBao replied:
Since EPS is scalar I suspect torch would down cast it when computing, if it is following the same casting logic for arithmetic regarding python number and torch.tensor. You might want to check out `aten/src/ATen/native/cpu/layer_norm_kernel.cpp`

titaiwangms replied:
Done

titaiwangms committed:
Update on "[ONNX] Fix ignored small eps in layer normalization in fp16"


Prior to this change, the symbolic_fn `layer_norm` (before ONNX version 17) always lose precision when eps is smaller than Float type, while PyTorch always take eps as Double. This PR adds `onnx::Cast` into eps related operations to prevent losing precision during the calculation.

[ghstack-poisoned]

titaiwangms commented:
@BowenBao PTAL

titaiwangms commented:
@pytorchbot merge -g
